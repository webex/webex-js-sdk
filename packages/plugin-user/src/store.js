/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 */

import {patterns} from '@ciscospark/common';

const usersByEmail = new WeakMap();
const usersById = new WeakMap();

export default class UserStore {
  constructor() {
    usersByEmail.set(this, new Map());
    usersById.set(this, new Map());
  }

  add(user) {
    if (!user.id) {
      return Promise.reject(`\`user.id\` is required`);
    }

    if (!user.emailAddress) {
      return Promise.reject(`\`user.emailAddress\` is required`);
    }

    const p1 = this.getById(user.id)
      .then((u) => usersById.get(this).set(user.id, Object.assign({}, u, user)))
      .catch(() => usersById.get(this).set(user.id, Object.assign({}, user)));

    const p2 = this.getByEmail(user.emailAddress)
      .then((u) => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, u, user)))
      .catch(() => usersByEmail.get(this).set(user.emailAddress, Object.assign({}, user)));

    return Promise.all([p1, p2]);
  }

  get(id) {
    if (patterns.uuid.test(id)) {
      return this.getById(id);
    }

    if (patterns.email.test(id)) {
      return this.getByEmail(id);
    }

    return Promise.reject(`\`id\` does not appear to be a valid user identifier`);
  }

  getById(id) {
    const ret = usersById.get(this).get(id);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }

  getByEmail(email) {
    const ret = usersByEmail.get(this).get(email);
    if (ret) {
      return Promise.resolve(ret);
    }
    return Promise.reject(new Error(`No user found identified by specified identifier`));
  }
}
