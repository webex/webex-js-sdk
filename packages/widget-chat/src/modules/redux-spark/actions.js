export const UPDATE_SPARK_STATE = `UPDATE_SPARK_STATE`;
export function updateSparkState(status) {
  return {
    type: UPDATE_SPARK_STATE,
    payload: {
      status
    }
  };
}

export const STORE_SPARK_INSTANCE = `STORE_SPARK_INSTANCE`;
export function storeSparkInstance(spark) {
  return {
    type: STORE_SPARK_INSTANCE,
    payload: {
      spark
    }
  };
}


export function registerDevice(spark) {
  return (dispatch) => {
    dispatch(updateSparkState({registering: true}));
    return spark.device.register()
      .then(() => dispatch(updateSparkState({registering: false, registered: true})));
  };
}

export function connectToMercury(spark) {
  return (dispatch) => {
    dispatch(updateSparkState({connecting: true}));
    return spark.mercury.connect()
    .then(() => dispatch(updateSparkState({connecting: false, connected: true})));
  };
}
