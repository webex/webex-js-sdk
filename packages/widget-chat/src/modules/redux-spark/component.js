import React, {Component} from 'react';
import {connect} from 'react-redux';

import {
  connectToMercury,
  updateSparkState
} from './actions';

class SparkComponent extends Component {

  componentDidMount() {
    const {
      spark
    } = this.props;

    spark.listenToAndRun(spark, `change:isAuthenticated`, () => {
      updateSparkState({authenticated: spark.isAuthenticated});
    });

    spark.listenToAndRun(spark, `change:isAuthenticating`, () => {
      updateSparkState({authenticating: spark.isAuthenticating});
    });

    spark.mercury.listenToAndRun(spark.mercury, `change:connected`, () => {
      updateSparkState({connected: spark.mercury.connected});
    });

    spark.mercury.listenToAndRun(spark.mercury, `change:connecting`, () => {
      updateSparkState({connecting: spark.mercury.connecting});
    });
  }

  componentWillReceiveProps(nextProps) {
    const {
      authenticated,
      authenticating,
      connected,
      connecting,
      spark
    } = nextProps;
    if (authenticated && !connected && !connecting && !authenticating) {
      connectToMercury(spark);
    }
  }

  render() {
    return null;
  }
}

SparkComponent.propTypes = {
  connectToMercury: React.PropTypes.func.isRequired,
  spark: React.PropTypes.object.isRequired,
  updateSparkState: React.PropTypes.func.isRequired
};

function mapStateToProps(state, ownProps) {
  return {
    connectToMercury,
    updateSparkState,
    spark: ownProps.spark
  };
}

export default connect(
  mapStateToProps
)(SparkComponent);
