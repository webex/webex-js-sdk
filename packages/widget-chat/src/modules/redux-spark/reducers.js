import {Map} from 'immutable';

import {
  STORE_SPARK_INSTANCE,
  UPDATE_SPARK_STATE
} from './actions';

export default function reduceSpark(state = new Map({
  status: new Map({
    authenticated: false,
    authenticating: false,
    registered: false,
    registering: false,
    connected: false,
    connecting: false
  }),
  spark: null
}), action) {
  switch (action.type) {

  case UPDATE_SPARK_STATE:
    return state.mergeDeep({
      status: action.payload.status
    });

  case STORE_SPARK_INSTANCE:
    return state.setIn([`spark`], action.payload.spark);

  default:
    return state;
  }
}
