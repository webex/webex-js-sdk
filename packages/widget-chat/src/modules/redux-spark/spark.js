/* eslint camelcase: [0] */

import '@ciscospark/plugin-conversation';
import Spark from '@ciscospark/spark-core';

export default function createSpark(accessToken) {
  const credentials = JSON.parse(localStorage.getItem(`credentials`) || false);
  return new Spark({
    credentials: {
      authorization: {
        access_token: accessToken || process.env.CISCOSPARK_ACCESS_TOKEN || credentials.access_token
      }
    }
  });
}
