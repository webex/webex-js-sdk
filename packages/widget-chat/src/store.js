import {
  applyMiddleware,
  combineReducers,
  compose,
  createStore
} from 'redux';

import thunk from 'redux-thunk';
import createLogger from 'redux-logger';
import {
  activity,
  conversation,
  flags,
  indicators,
  notifications,
  share,
  user,
  widget
} from './reducers';
import sparkReducer from './modules/redux-spark/reducers';

const devtools = window.devToolsExtension || (() => (noop) => noop);

const enhancers = [
  applyMiddleware(
    thunk,
    createLogger({
      level: `info`,
      duration: true,
      collapsed: false,
      logErrors: process.env.NODE_ENV === `production`
    })
  ),
  devtools()
];

export default createStore(
  combineReducers({
    activity,
    conversation,
    flags,
    indicators,
    notifications,
    share,
    user,
    widget,
    spark: sparkReducer
  }),
  compose(...enhancers)
);
