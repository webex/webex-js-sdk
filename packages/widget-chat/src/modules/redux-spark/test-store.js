// Store for testing
import {
  applyMiddleware,
  combineReducers,
  compose,
  createStore
} from 'redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {Map} from 'immutable';
import reducers from './reducers';

jest.mock(`./spark`);
const {createSpark} = require(`./spark`);

export function createMockStore() {
  const mockStore = configureMockStore([thunk]);
  return mockStore(new Map({
    status: new Map({
      authenticated: false,
      authenticating: false,
      registered: false,
      registering: false,
      connected: false,
      connecting: false
    }),
    spark: createSpark()
  }));
}

const enhancers = [
  applyMiddleware(
    thunk)
];

export const store = createStore(
  combineReducers({
    spark: reducers
  }),
  compose(...enhancers)
);
