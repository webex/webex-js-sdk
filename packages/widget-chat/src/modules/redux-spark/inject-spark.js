import React, {Component} from 'react';
import spark from './spark';

import SparkComponent from './component.js';

function getDisplayName(C) {
  return C.displayName || C.name || `C`;
}

export default function injectSpark(WrappedComponent, options = {}) {
  const {
    withRef = false
  } = options;

  class InjectSpark extends Component {
    getWrappedInstance() {
      // TODO find a way to do this that doesn't require a deprecated API
      // eslint-disable-next-line react/no-string-refs
      return this.refs.wrappedInstance;
    }

    render() {
      return (
        <div>
          <SparkComponent spark={spark} />
          <WrappedComponent
            {...this.props}
            ref={withRef ? `wrappedInstance` : null}
            spark={spark}
          />
        </div>
      );
    }
  }

  InjectSpark.displayName = `InjectSpark(${getDisplayName(WrappedComponent)})`;
  InjectSpark.WrappedComponent = WrappedComponent;

  return InjectSpark;
}
