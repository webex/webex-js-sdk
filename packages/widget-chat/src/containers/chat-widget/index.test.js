import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';

import {store as sparkStore} from '../../modules/redux-spark/test-store';
import store from '../../store';
import ChatWidget from '.';

store.spark = sparkStore;

describe(`ChatWidget`, () => {
  it(`renders properly`, () => {
    const accessToken = process.env.CISCOSPARK_ACCESS_TOKEN;
    const widget = renderer.create(
      <Provider store={store}>
        <ChatWidget accessToken={accessToken} userId="bernie.zang@gmail.com" />
      </Provider>
    ).toJSON();

    expect(widget).toMatchSnapshot();
  });
});
