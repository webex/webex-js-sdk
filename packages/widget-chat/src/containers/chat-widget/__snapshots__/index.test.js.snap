exports[`ChatWidget renders properly 1`] = `
<div>
  <div
    className="widget-chat">
    <div
      className="banner" />
    <div
      className="loading">
      Connecting...
      <div
        className="spinner-container">
        <div
          className="spinner" />
      </div>
    </div>
  </div>
</div>
`;
